assembly/src/assembler/span_builder.rs

```diff
@@ -102,8 +102,11 @@ impl SpanBuilder {
     ///
     /// This indicates that the provided instruction should be tracked and the cycle count for
     /// this instruction will be computed when the call to set_instruction_cycle_count() is made.
-    pub fn track_instruction(&mut self, instruction: &Instruction) {
-        let op = AssemblyOp::new(instruction.to_string(), 0);
+    pub fn track_instruction(&mut self, instruction: &Instruction, ctx: &mut AssemblyContext) {
+        let proc = ctx.current_context_name();
+        let op = instruction.to_string();
+        let op = format!("{proc}:{op}");
+        let op = AssemblyOp::new(op, 0);
```

Comment:
I like this change a lot! But I wonder, why not add more fields to `AssemblyOp` rather than tracking additional info in the string?

For example, we define `AssemblyOp` as follows:
```Rust
pub struct AssemblyOp {
    op: String,
    ctx_name: String,
    should_break: bool,
    num_cycles: u8,
}
```
And then we:
- Don't need to do any string matching down the road.
- Don't need to make `processor` depend on `assembler` (which I'd like to avoid).
- Maybe be able to convert `breakpoint` instruction into a pure decorator (but I'm not 100% sure on this).

Reply:
That makes sense. To be fair, I think it's better we remove the new instruction entirely and make a breakpoint based on the decorator. Something like `break on "foo:swap"` where `foo` is the procedure, and `swap` is an instruction

The downside of this approach is we will stop at every incidence of `swap` inside `foo`, but we have the big benefit of not having to touch delicate questions such as changing the program root to make debugging feasible.

wdyt @grjte @hackaugusto ?

Also, I'm not sure if it's possible to use this breakpoint as pure decorator. The vm state iterates per tick, and a tick is an operation. The way I see, we need to have an operation there (in this case, `noop`).

Reply:
If we'd rather stick to the masm change to break, then a better solution would be to add a directive to an operation, and this directive would interact with the breakpoint. Something like:

```
#[break]
dup.1
```

Another way of doing this is make the `Breakpoint` affect the next op. Something like this

```
breakpoint
dup.1
```

Then we insert the breakpoint as decorator of `dup.1`. But that change will break the current model of one instruction per compiled operation.

I would favor going for the directive, as this can be useful for other things in the future too. The bad part is this is a new feature.

Reply:
My preference order on how to go for this:

1) naive break on `{procedure}:{instruction}` - will be closer to the final target, after we have source mapping
2) directive `#[break] dup.1` - will be able to extend for future functionality of the decorator
3) keep as is now
4) instruction affects the decorator of the next instruction - will break the current parsing model

Reply:
My thinking these is as follows:

1. I think `{procedure}:{instruction}` may not be granular enough as we have some procedures which are pretty long and we might have to go through dozens (or even hundreds) of breaks before we get to something we care about. 
2. I'd prefer to not to introduce new syntax.
3. I think keeping things as they are now (but we should add extra info to `AssemblyOp` instead of doing string matching).
4. This would actually be pretty simple (would not require changes to parsing and would not introduce new decorators) - but it won't always work as sometimes there would be no "next" instruction (i.e., "end" is not really an instruction). If we are fine saying that you can put a breakpoint right before "end", and I think this might be the simplest approach.

Reply:
I'd lean towards keeping it as it is now but with the minor change to adjust `AssemblyOp` rather than making more complex changes. It's a fairly small, straightforward change that works & gives a significant improvement in functionality. I think that `{procedure}:{instruction}` provides a worse debugging experience, sometimes significantly so. I also agree about not adding new syntax, since we plan to switch to source mapping later anyway.

I just want to clarify about your response @bobbinth:
1. Are you saying you prefer option 4 or option 3?

> If we are fine saying that you can put a breakpoint right before "end"

2. Did you mean saying that you _can't_ put a breakpoint right before "end"? It may be annoying, as this sort of thing can be a bit hard to communicate in a UX-friendly way, so people will probably only discover it when they try to break at the end of a procedure or program to view the state. That said, I do think it would be ok to do this, since it's still much better than what we have. In that case, is this your preference ahead of option 3?

Reply:
For simplicity, I'd go with option 3 for now.